### ISSUE TITLE
Path template matching needs more specific matching rules according to verb matching

### ISSUE THREAD

**Issue by wsalembi** [OAI/OpenAPI-Specification#2343]
The path template matching rules in the spec lead to confusion in several implementations of OpenAPI validators.
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#path-templating-matching

Given the specification https://petstore3.swagger.io. What is the right result when a client sends a request 
curl -X POST "https://petstore3.swagger.io/api/v3/pet/findByStatus"

Should this call be matched with path "/pet/findByStatus" and then result in a 405 Method Not Allowed error since the path has no POST verb.

Or can this call be matched with path  "/pet/{petId}" and result in a call to operationId updatePetWithForm?

Per specification: When matching URLs, concrete (non-templated) paths would be matched before their templated counterparts. Templated paths with the same hierarchy but different templated names MUST NOT exist as they are identical. In case of ambiguous matching, it's up to the tooling to decide which one to use.

So per spec: POST to /pet/findByStatus should match the path /pet/findByStatus since it's an exact match of a concrete path.

What should vendors to if paths match, but verbs don't? Continue matching with other paths or abort?

**Comment by handrews** [comment]
TL;DR: path templating has to ignore whether a method is present, because there may be an actual implemented method that does not show up in the OpenAPI Description.

@wsalembi This got lost in our years of epic issue backlog, and somehow I didn't notice it when we cleared out a lot of that backlog last year.  If you are still paying attention, the OAS describes things that _can_ be done, but in many cases does _not_ describe what _can't_ be done.  So the lack of a `post` field under matching path just means that _if_ there is a POST operation on that endpoint, we have no idea what it does, and nothing about it can be automated by OpenAPI tooling.

So we would definitely _not_ match it to `/pet/{petId}`, because the matching really does need to be independent of HTTP method in case there is an undescribed method present.  This may seem odd, but the method may only be usable by some API clients (e.g. you will get an auth error, and the method was filtered out of the version of the OpenAPI Description that you are allowed to see).

**Comment by karenetheridge** [comment]
Don't forget that there's more to the matching than just the paths template -- the request URI must match against the path template and a server url as well. Therefore we cannot know that this particular path entry is even a match without having a matching method, because server urls are used as part of the matching, and server can come from the operation object, which depends on the method existing in the path-item object.

The specification (quoting from the 3.1.1 version of the doc) says in §4.8.8.1: "In case of ambiguous matching, it’s up to the tooling to decide which one to use." I suspect the case described by @wsalembi would count as an ambiguous match (where more than one entry might do?  In any case, if this is _not_ the intent, it could do with some clarification in the spec.

**Comment by karenetheridge** [comment]
I've run into several more ambiguous match scenarios where one simply has to go with _something_ that seems to make sense for the usecases involved.  I've documented some of these, with the advice that if there is ever any possibility of confusion when making a match, including the operationId with the match request is advisable, as this removes all ambiguity with the match and lets the implementation get right down to validating everything else about the request against what's specified in the path-item and operation objects.

I don't see how it's helpful for the spec to insist that a lack of a match could simply mean that the API is not documented.  In some usecases I could see the desired behaviour to be something else - a 404 response when the path-item is not found, a 405 if the path-item exists but there is no matching method, and so on. \[However I would assume that when deployed in an actual webserver, some other URI matching/dispatching system is being used ahead of the OpenAPI validation, so the matching part of the implementation is not as crucial.\]

**Comment by handrews** [comment]
@karenetheridge this got auto-closed and I can't tell if any of the other issues/discussions on this topic cover what you mentioned.  Feel free to re-open or consolidate or whatever.  I feel like a lot of the path templating / request matching issues overlap and could be consolidated down to something more clear, but I'm not familiar enough with this area to do that.

**Comment by karenetheridge** [comment]
> path templating has to ignore whether a method is present, because there may be an actual implemented method that does not show up in the OpenAPI Description.

If the API document is to "fail open", as in presume that there might be other APIs that exist that are undocumented, this should be clarified in the spec. The concept of "fail open" vs "fail closed" came up at the Moonwalk discussion yesterday and we considered the possibility that this could be a configuration value at the top level of the OpenAPI Document (OAD).

I think we can keep this issue closed and consolidate the various points into a discussion in the moonwalk repository.
